lib/openshift/ingress_controller.rb

```diff
@@ -0,0 +1,24 @@
+module BushSlicer
+  class IngressController < ProjectResource
+    RESOURCE = "ingresscontrollers"
+
+
+    ### status related methods
+    def available_replicas(user: nil, cached: true, quiet: false)
+      raw_resource(user: user, cached: cached, quiet: quiet).dig('status', 'availableReplicas')
+    end
```

Comment:
Can this be a schild of `PodReplicator` like `Deployment` and have this method inheritted?

Reply:
I thought about it, but ingress and pods are two very different kind of things...it just didn't sound right to me.

Reply:
`ingresscontrollers.operator.openshift.io` is a CustomResource, maybe it should be a child of `ClusterResource`?

Reply:
Custom vs non-custom does not affect whether resource is namespaced or not [1] .

If ingress does not replicate pods, then you are right. If replicas are pods, then it must be `PodReplicator`. What are these replicas?


[1] https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/#not-all-objects-are-in-a-namespace

Reply:
Basically if `PodReplicator` methods can be implemented for ingress, then can be a subclass.

Reply:
Yes, `ingresscontroller` can replicate pods, the operator read the spec.replicas and update `deployment`.

Reply:
@lihongan thanks for the tips.  @akostadinov I'll change it to inherit from `PodReplicator` then.

Reply:
Also method `available_replicas` needs to be deleted and contant REPLICA_COUNTERS to be added instead. See `lib/openshift/deployment.rb` for example. 

Make sure that `#template` and `#containers_spec` methods from `PodReplicator` also make sense for this class. If they do NOT, then this is NOT a `PodReplicator` after all.

Reply:
@akostadinov hmm...it looks like we can't inherit it from PodReplicator as indicated below, there are no `template` nor `container_spec`

```apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
  creationTimestamp: "2019-05-28T17:04:03Z"
  finalizers:
  - ingresscontroller.operator.openshift.io/finalizer-ingresscontroller
  generation: 1
  name: default
  namespace: openshift-ingress-operator
  resourceVersion: "14383"
  selfLink: /apis/operator.openshift.io/v1/namespaces/openshift-ingress-operator/ingresscontrollers/default
  uid: 98af0220-816a-11e9-82c1-02b3adb00b00
spec: {}
status:
  availableReplicas: 2
  conditions:
  - lastTransitionTime: "2019-05-28T17:07:59Z"
    status: "True"
    type: Available
  domain: apps.qe-pruan-05-28.qe.devcluster.openshift.com
  endpointPublishingStrategy:
    type: LoadBalancerService
  selector: ingresscontroller.operator.openshift.io/deployment-ingresscontroller=default```

Reply:
@akostadinov should we put it back to what I had originally then?

Reply:
yes, but we may create a class `Replicable` that only handles the `replicas` stuff. Then have `PodReplicator` inherit from `Replicable` and `IngressController` also inherit from `Replicable`. Is this explanation good enough?

Reply:
Do we have a solution for this?
FYI, the `spec` is null in the default IngressController, but in a custom IngressController it might be like below:
```
spec:
  defaultCertificate:
    name: test-certs
  domain: router-test.example.com
  replicas: 1
  nodePlacement:
    nodeSelector:
      matchLabels:
        beta.kubernetes.io/os: linux
    tolerations:
    - effect: NoSchedule
      operator: Exists
  routeSelector:
    matchLabels:
      route: router-test
  namespaceSelector:
    matchLabels:
      namespace: router-test
```

Reply:
It should return `0` when `spec` is `nil`. So I think it would work fine.

Reply:
The operator will set default (hard coded) value for some parameters when `spec` is nil. 
For example, replicas will be set to 2 if not defined.
```
spec: {}
status:
  availableReplicas: 2
```

Reply:
Then `availableReplicas` should be read from `status`, not `spec`. I don't think this is an issue.